### PR TITLE
Extract holidays into a separate collection, add DateTimePeriod for i…

### LINKED ISSUE #6: Accept DatePeriods as holidays
Holidays are usually not just one day (especially in summer). DatePeriods can be iterated through, so you can transform them into days internally.

But actually I wonder if holidays could be some kind of special day as well. For example a Holiday object. It could be used to ease creating different types of holidays (simple DateTime or DatePeriod).

If you like any of the above, I would be happy to provide a PR.

Big :+1: for this package. I was thinking about one for years now, but never had the time to make it.

### PR DESCRIPTION
**Pull request by sagikazarmark**
…ntervals

Close #6

### CODE REVIEW

File: src/Business.php [hunk 1]
```diff
@@ -26,13 +26,13 @@
      * Creates a new business.
      *
      * @param DayInterface[]     $days
-     * @param \DateTime[]        $holidays
+     * @param Holidays|null      $holidays
      * @param \DateTimeZone|null $timezone
      */
-    public function __construct(array $days, array $holidays = [], \DateTimeZone $timezone = null)
+    public function __construct(array $days, Holidays $holidays = null, \DateTimeZone $timezone = null)
```
Comment (florianv) [inline src/Business.php]: The typehint change of `$holidays` is a BC break

File: src/DateTimePeriod.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+{
+    private $datePeriod;
+
+    /**
+     * Creates a new DateTime period.
+     *
+     * @param \DateTime $startDate
+     * @param \DateTime $endDate
+     */
+    public function __construct(\DateTime $startDate, \DateTime $endDate)
```
Comment (florianv) [inline src/DateTimePeriod.php]: You need to check that $endDate is after $startDate

File: src/DateTimePeriod.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        // We need to add one day to include the end date as well
```
Comment (florianv) [inline src/DateTimePeriod.php]: I would remove this comment

File: src/DateTimePeriod.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        $endDate = $endDate->modify('+1 day');
+
+        $this->datePeriod = new \DatePeriod($startDate, new \DateInterval('P1D'), $endDate);
+    }
+
+    public function getIterator()
```
Comment (florianv) [inline src/DateTimePeriod.php]: Missing @inheritdoc comment

File: src/DateTimePeriod.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * Creates a new DateTime period.
+     *
+     * @param \DateTime $startDate
+     * @param \DateTime $endDate
+     */
+    public function __construct(\DateTime $startDate, \DateTime $endDate)
+    {
+        // We need to add one day to include the end date as well
+        $endDate = $endDate->modify('+1 day');
```
Comment (florianv) [inline src/DateTimePeriod.php]: $endDate needs to be cloned otherwise the constructor will modify it (it's not immutable)

Comment (florianv) [inline src/DateTimePeriod.php]: The `=` is not needed because `modify()` directly modifies the instance

File: src/Holidays.php [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+ * (c) Florian Voutzinos <florian@voutzinos.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Business;
+
+/**
+ * Default implementation of BusinessInterface.
```
Comment (florianv) [inline src/Holidays.php]: Wrong comment

File: src/Holidays.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function unserialize($serialized)
+    {
+        $this->holidays = unserialize($serialized);
+    }
+
+    private function addHoliday(\DateTime $holiday)
```
Comment (florianv) [inline src/Holidays.php]: Please put the private methods after all the public methods

**Comment by florianv** [comment]
Thanks for the PR!

Looks good once the commented issues are fixed

File: src/Business.php (hunk 1 above)
Comment (sagikazarmark) [inline src/Business.php]: Indeed. How should I proceed with the type check?

If an array is passed it is wrapped in a Holidays object. If null is passed, an empty object is created. What about other types? Check in the Business constructor and throw an invalid argument exception?

Comment (florianv) [inline src/Business.php]: Yes

**Comment by sagikazarmark** [comment]
Pushed the required changes. If you wish, I can squash commits after review and rebase #9 after merge.

File: src/Holidays.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+     *
+     * @param \DateTime[]|DateTimePeriod[] $holidays
+     */
+    public function __construct(array $holidays = [])
+    {
+        $this->addHolidays($holidays);
+    }
+
+    /**
+     * Checks if a given date is holiday
```
Comment (florianv) [inline src/Holidays.php]: Missing `.` at the end of the sentence

File: src/Holidays.php [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+     *
+     * @param \DateTime $date
+     *
+     * @return boolean
+     *
+     * @internal
```
Comment (florianv) [inline src/Holidays.php]: :+1:

File: src/Holidays.php [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        $this->addHolidays($holidays);
+    }
+
+    /**
+     * Checks if a given date is holiday
+     *
+     * @param \DateTime $date
+     *
+     * @return boolean
```
Comment (florianv) [inline src/Holidays.php]: Please use `bool` instead (in order to match HHVM and PHP 7 type names)

**Comment by florianv** [comment]
Nice work :+1: Looks good for me once the small issues are fixed.

If you could add an example to the documentation about how to use the `DateTimePeriod ` using `Holidays` it would be perfect :)

> If you wish, I can squash commits after review

Yes, thanks.

**Comment by sagikazarmark** [comment]
Issues fixed. Please review again. After that I squash commits.

File: src/DateTimePeriod.php [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+     * @param \DateTime $startDate
+     * @param \DateTime $endDate
+     */
+    public function __construct(\DateTime $startDate, \DateTime $endDate)
+    {
+        $endDate = clone $endDate;
+        $endDate->modify('+1 day');
+
+        if ($startDate >= $endDate) {
+            throw new \InvalidArgumentException('Start date must be earlier than end date.');
```
Comment (sagikazarmark) [inline src/DateTimePeriod.php]: In `Business` object `LogicException` is thrown for the same case. Should I throw that one here as well?

Comment (florianv) [inline src/DateTimePeriod.php]: Yes please

File: README.md [hunk 12]
```diff
@@ -74,6 +80,7 @@ $nextDate = $business->closest(new \DateTime('2015-05-11 10:00'));
 $lastDate = $business->closest(new \DateTime('2015-05-11 10:00'), Business::CLOSEST_LAST);
 ```
 
+
```
Comment (florianv) [inline README.md]: This extra line can be removed

**Comment by sagikazarmark** [comment]
Changes and squash done

**Comment by florianv** [comment]
Thanks @sagikazarmark

**Pull request merged**